src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/DynamicDependencyAttributeAlgorithm.cs

```diff
@@ -26,135 +30,187 @@ public static void AddDependenciesDueToDynamicDependencyAttribute(ref Dependency
 
-                if (fixedArgs.Length > 0 && fixedArgs[0].Value is string sigFromAttribute)
+            var fixedArgs = attribute.FixedArguments;
+            TypeDesc targetType;
+
+            if (fixedArgs.Length > 0 && fixedArgs[0].Value is string sigFromAttribute)
+            {
+                if (fixedArgs.Length == 1)
                 {
-                    if (fixedArgs.Length == 1)
-                    {
-                        // DynamicDependencyAttribute(String)
-                        targetType = method.OwningType;
-                    }
-                    else if (fixedArgs.Length == 2 && fixedArgs[1].Value is TypeDesc typeFromAttribute)
-                    {
-                        // DynamicDependencyAttribute(String, Type)
-                        targetType = typeFromAttribute;
-                    }
-                    else if (fixedArgs.Length == 3 && fixedArgs[1].Value is string typeStringFromAttribute
-                        && fixedArgs[2].Value is string assemblyStringFromAttribute)
+                    // DynamicDependencyAttribute(String)
+                    targetType = owningType;
+                }
+                else if (fixedArgs.Length == 2 && fixedArgs[1].Value is TypeDesc typeFromAttribute)
+                {
+                    // DynamicDependencyAttribute(String, Type)
+                    targetType = typeFromAttribute;
+                }
+                else if (fixedArgs.Length == 3 && fixedArgs[1].Value is string typeStringFromAttribute
+                    && fixedArgs[2].Value is string assemblyStringFromAttribute)
+                {
+                    // DynamicDependencyAttribute(String, String, String)
+                    ModuleDesc asm = factory.TypeSystemContext.ResolveAssembly(new System.Reflection.AssemblyName(assemblyStringFromAttribute), throwIfNotFound: false);
+                    if (asm == null)
                     {
-                        // DynamicDependencyAttribute(String, String, String)
-                        ModuleDesc asm = factory.TypeSystemContext.ResolveAssembly(new System.Reflection.AssemblyName(assemblyStringFromAttribute), throwIfNotFound: false);
-                        if (asm == null)
-                        {
-                            // _context.LogWarning($"Unresolved assembly '{dynamicDependency.AssemblyName}' in 'DynamicDependencyAttribute'", 2035, context);
-                            continue;
-                        }
-
-                        targetType = DocumentationSignatureParser.GetTypeByDocumentationSignature((IAssemblyDesc)asm, typeStringFromAttribute);
-                        if (targetType == null)
-                        {
-                            // _context.LogWarning ($"Unresolved type '{typeName}' in DynamicDependencyAttribute", 2036, context);
-                            continue;
-                        }
+                        (factory.MetadataManager as UsageBasedMetadataManager)?.Logger?.LogWarning(
```

Comment:
Is the `is` needed? Basically if we run this with other than the UsageBased manager, we're running this logic twice (first time scanner, second time real compilation). We should ideally not do that - it's useless work the second time.

```suggestion
                        ((UsageBasedMetadataManager)factory.MetadataManager).Logger.LogWarning(
```

Reply:
Changed